File: python_practice/202.快乐数.py
#
# @lc app=leetcode.cn id=202 lang=python3
#
# [202] 快乐数
#


# @lc code=start
class Solution:
    def isHappy(self, n: int) -> bool:
        def cal_sum(n: int) -> int:
            '''calculate helper'''
            sum = 0
            while n > 0:
                last = n % 10
                n = n // 10
                sum += last * last
            return sum

        history = set()
        while True:
            if n == 1:
                return True
            else:
                history.add(n)
                n = cal_sum(n)
                if n in history:
                    return False


# @lc code=end

File: cpp/202.快乐数.cpp
/*
 * @lc app=leetcode.cn id=202 lang=cpp
 *
 * [202] 快乐数
 */
#include <unordered_set>
using namespace std;
// @lc code=start
class Solution {
public:
    bool isHappy(int n) {
        unordered_set<int> history;
        while (true)
        {
            int sum = getSum(n);
            if (sum == 1) {
                return true;
            }
            if (history.find(sum) != history.end()) {
                return false;
            } else {
                n = sum;
                history.insert(n);
            }
        }
    }

    int getSum(int n) {
        int sum = 0;
        while (n > 0)
        {
            int last = n % 10;
            n = n / 10;
            sum += last * last;
        }
        return sum;
    }
};
// @lc code=end

